### PR TITLE
Re-adding execvpe support, conditional on platform

### LINKED ISSUE #682: Add execvpe support
This was removed in #681 because it was broken. Re-adding it is basically undoing fcc22f22a680233dbeb82d230e35fd54cb6de176 (without the feature support and instead conditionally including it on supported platforms). `libc` has `execvpe` included for `haiku`, `linux`, and `openbsd` platforms. If you are a new contributor, go ahead and read `CONTRIBUTING.md`, assign this to yourself, and file a PR when you've gotten somewhere (start the title with `WIP` if it's not ready to merge, and include "closes #682" in the text of the PR).

### PR DESCRIPTION
**Pull request by panzertime**
Fixes #682

### CODE REVIEW

**Comment by Susurrus** [comment]
You'll need to do a few more things:
  * replace `super::ffi::execvpe` with `libc::execvpe`
  * Change the last line to use `Error::result` like in the function above this one.
  * There should be good documentation for this function. See some of the other PRs that were recently merged for good examples.

File: src/unistd.rs [hunk 1]
```diff
@@ -1644,6 +1644,15 @@ mod linux {
         Errno::result(res).map(drop)
     }
 
+    /// Replaces the current process image with a new one.
+    /// ([see exec(3)](http://man7.org/linux/man-pages/man3/exec.3.html))
```
Comment (Susurrus) [inline src/unistd.rs]: You need a blank line below this for the docs to look good. Please run `cargo doc --no-deps --open` and look at the `unistd` module to see the summary and click on `execvpe` to view the full description. This doesn't render well in the docs.

File: src/unistd.rs [hunk 2]
```diff
@@ -1644,6 +1644,15 @@ mod linux {
...
+    /// * `filename`: filename of executable to run
+    /// * `args`: array of arguments as strings; first must be the filename
+    /// * `env`: array of environment variables for the new execution environment
+    /// * returns: Ok or libc error code.
+    ///
+    /// Err is returned if the executable file does not exist, cannot be accessed, or
```
Comment (Susurrus) [inline src/unistd.rs]: Instead of "Err" use "An error"

File: src/unistd.rs [hunk 3]
```diff
@@ -1656,10 +1665,10 @@ mod linux {
...
         env_p.push(ptr::null());
 
-        unsafe {
-            super::ffi::execvpe(filename.as_ptr(), args_p.as_ptr(), env_p.as_ptr())
+        let res = unsafe {
+            libc::execvpe(filename.as_ptr(), args_p.as_ptr(), env_p.as_ptr())
         };
 
-        Err(Error::Sys(Errno::last()))
+	Errno::result(res).map(drop)
```
Comment (Susurrus) [inline src/unistd.rs]: Indentation here is off by 1.

**Comment by panzertime** [comment]
I have made all the requested changes, but I cannot squash all the changes into two commits as described

File: src/unistd.rs [hunk 4]
```diff
@@ -1644,8 +1644,18 @@ mod linux {
         Errno::result(res).map(drop)
     }
 
+    /// Replaces the current process image with a new one.
+    /// ([see exec(3)](http://man7.org/linux/man-pages/man3/exec.3.html))
+    ///
+    /// * `filename`: filename of executable to run
```
Comment (Susurrus) [inline src/unistd.rs]: Can you give a title or quick sentence to introduce this bulleted list? It doesn't read nicely just being a stand alone list.

File: src/unistd.rs [hunk 5]
```diff
@@ -1644,8 +1644,18 @@ mod linux {
...
+    /// * `args`: array of arguments as strings; first must be the filename
+    /// * `env`: array of environment variables for the new execution environment
+    /// * returns: Ok or libc error code.
+    ///
+    /// An error is returned if the executable file does not exist, cannot be accessed, or
```
Comment (Susurrus) [inline src/unistd.rs]: There is a special section called `Errors` that you should use for these kinds of things. See [the book](https://doc.rust-lang.org/book/first-edition/documentation.html#special-sections).

File: src/unistd.rs [hunk 6]
```diff
@@ -1644,8 +1644,18 @@ mod linux {
         Errno::result(res).map(drop)
     }
 
+    /// Replaces the current process image with a new one.
+    /// ([see exec(3)](http://man7.org/linux/man-pages/man3/exec.3.html))
+    ///
+    /// * `filename`: filename of executable to run
+    /// * `args`: array of arguments as strings; first must be the filename
+    /// * `env`: array of environment variables for the new execution environment
+    /// * returns: Ok or libc error code.
```
Comment (Susurrus) [inline src/unistd.rs]: This can be handled by the `Errors` section below.

**Comment by Susurrus** [comment]
I would like to get these commits squashed, you said you couldn't do that, is that because you aren't familiar with `git rebase`? [Here's a quick overview](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html) if that's all you're missing.

**Comment by panzertime** [comment]
I followed the exact procedure described there and the rebase failed.

**Comment by Susurrus** [comment]
You'll need to fix merge commits when you rebase sometimes, so if there's a conflict it isn't a "failure" it just means you need to manually merge some of the code.

**Comment by panzertime** [comment]
Well, if that's the case, someone else needs to do it.

**Comment by Susurrus** [comment]
This isn't that bad to fix and it's probably worth getting a little bit more familiar with `git` if you're going to be doing more software dev work. And I've been where you are and had to struggle through it a bit, but the solution to this is pretty easy: undo all of your commits locally, rebase to the latest upstream/master (I'm assuming the primary repo is your upstream here), and then make your two commits.

This looks like:
  1) `git fetch upstream` - Make sure you have the latest nix code to reference
  2) `git reset upstream/master` - This will keep all your changes but unstage them and undo all of the commits. You'll now be at the same commit as upstream/master but with none of your changes applied.
  3) `git commit CHANGELOG.md -m"Switch back to using libc git"`
  4) `git commit -a -m"Re-add execvpe for Haiku, OpenBSD, and Linux"
  5) `git push origin -f` - Overwrite the remote branch. Since this replaces existing commits you need to use the `-f`. Thinking about this now, I'm wondering if this isn't the failure you were talking about above.

Now, if more changes need to be made:
  1) `git commit -a --amend` - Just amend the existing commit with any new changes.
  2) `git push origin -f` - Since you'll have rewritten the git history, you'll need to force-push again here.

Let me know if you get stuck doing any of that. Hopefully it all makes sense now!

**Comment by panzertime** [comment]
fatal: 'upstream' does not appear to be a git repository

**Comment by Susurrus** [comment]
You may have it called "origin"; do `git remote -v` to see the names of things. I have "origin" set to my fork of a repository and I use "upstream" to refer to the upstream repo.

**Comment by panzertime** [comment]
There was no remote set for the upstream repo.

**Comment by Susurrus** [comment]
Yeah, the default would be to have "origin" set to your repo. So let's add an upstream:

 * `git remote add upstream https://github.com/nix-rust/nix`
 * `git fetch upstream`

**Comment by panzertime** [comment]
I already did that.

**Comment by Susurrus** [comment]
What's the output of `git remote -v` for you then?

**Comment by panzertime** [comment]
origin	https://github.com/panzertime/nix.git (fetch)
origin	https://github.com/panzertime/nix.git (push)
upstream	https://github.com/nix-rust/nix/ (fetch)
upstream	https://github.com/nix-rust/nix/ (push)

**Comment by Susurrus** [comment]
Oh, perfect, I did not. I've been squatting on this page and GitHub's terrible at dealing with force-pushing things, so I was still seeing 12 commits or whatever.

So this wraps in some changes that are part of the commits in #681 into your own commits, which isn't ideal, but fine. What we'll do is wait until that's merged, rebase this (which will get rid of all of those code changes) and then this'll be ready to merge!

Thanks for sticking with this, git can be really frustrating sometimes, but I've found it quite powerful now that I've really gotten the hang of it.

**Comment by Susurrus** [comment]
@panzertime Want to rerebase this and clean up the remaining issues? This is pretty close to merging.

**Comment by panzertime** [comment]
What issues remain?

File: CHANGELOG.md [hunk 7]
```diff
@@ -24,6 +24,8 @@ This project adheres to [Semantic Versioning](http://semver.org/).
   and nix::Error::UnsupportedOperation}`
   ([#614](https://github.com/nix-rust/nix/pull/614))
 - Added `cfmakeraw`, `cfsetspeed`, and `tcgetsid`. ([#527](https://github.com/nix-rust/nix/pull/527))
+- Readded execvpe support, conditional on platform.  libc now provides this for Haiku, Linux, and OpenBSD.
```
Comment (Susurrus) [inline CHANGELOG.md]: Change to "Added `unistd::execvpe` for Haiku, Linux, and OpenBSD."

**Comment by Susurrus** [comment]
Just the CHANGELOG entry should be more clear & concise. Otherwise this needs a rebase and then it should be good. Though I want to take another look over it once it's rebased as it's hard to tell what the actual changes are here since there are some other things that seem mixed in here.

**Comment by Susurrus** [comment]
@panzertime Just pinging you on this since we were getting pretty close to merging this. Can you rebase and address my comments?

**Comment by Susurrus** [comment]
@panzertime It's been a month since I last pinged you, so I wanted to do that again and see if we could get this merged.

**Comment by Susurrus** [comment]
@panzertime I think it's just the changelog entry now and a rebase and this is good to go. Still interested in merging this?

**Comment by Susurrus** [comment]
It's been almost 5 months on this with several pings, so I'm going to go ahead and close this PR.

**Pull request closed without merging**